charmhub/client.go

```diff
@@ -19,170 +19,61 @@ import (
-	metadataHeaders map[string]string
-	transportFunc   func(Logger) Transport
-}
-
-// WithURL sets the url on the option.
-func WithURL(u string) Option {
-	return func(options *options) {
-		options.url = &u
-	}
-}
-
-// WithMetadataHeaders sets the headers on the option.
-func WithMetadataHeaders(h map[string]string) Option {
-	return func(options *options) {
-		options.metadataHeaders = h
-	}
-}
-
-// WithHTTPTransport sets the default http transport to use on the option.
-func WithHTTPTransport(transportFn func(Logger) Transport) Option {
-	return func(options *options) {
-		options.transportFunc = transportFn
-	}
-}
-
-// Create a options instance with default values.
-func newOptions() *options {
-	u := CharmHubServerURL
-	return &options{
-		url: &u,
-		transportFunc: func(logger Logger) Transport {
-			return DefaultHTTPTransport(logger)
-		},
-	}
-}
 
-// CharmHubConfig defines a charmHub client configuration for targeting the
-// charmhub API.
-func CharmHubConfig(logger Logger, options ...Option) (Config, error) {
-	opts := newOptions()
-	for _, option := range options {
-		option(opts)
-	}
-
-	if logger == nil {
-		return Config{}, errors.NotValidf("nil logger")
-	}
-
-	// By default we want to specify a default user-agent here. In the future
-	// we should ensure this probably contains model UUID and cloud.
-	headers := make(http.Header)
-	headers.Set(userAgentKey, userAgentValue)
-
-	// Additionally apply any metadata headers to the headers so we can send
-	// every time we make a request.
-	m := opts.metadataHeaders
-	keys := make([]string, 0, len(m))
-	for k := range m {
-		keys = append(keys, k)
-	}
-	sort.Strings(keys)
-	for _, k := range keys {
-		headers.Add(MetadataHeader, k+"="+m[k])
-	}
-
-	chLogger := logger.ChildWithLabels("client", corelogger.CHARMHUB)
-	return Config{
-		URL:       *opts.url,
-		Version:   CharmHubServerVersion,
-		Entity:    CharmHubServerEntity,
-		Headers:   headers,
-		Transport: opts.transportFunc(chLogger.ChildWithLabels("transport", corelogger.HTTP)),
-		Logger:    chLogger,
-	}, nil
+	// FileSystem represents the file system operations for downloading.
+	// If nil, use the real OS file system.
+	FileSystem FileSystem
 }
 
-// CharmHubConfigFromURL defines a charmHub client configuration with a given
-// URL for targeting the API.
-func CharmHubConfigFromURL(url string, logger Logger, options ...Option) (Config, error) {
-	return CharmHubConfig(logger, append(options, WithURL(url))...)
-}
-
-// BasePath returns the base configuration path for speaking to the server API.
-func (c Config) BasePath() (charmhubpath.Path, error) {
-	baseURL := strings.TrimRight(c.URL, "/")
-	rawURL := fmt.Sprintf("%s/%s", baseURL, path.Join(c.Version, c.Entity))
+// basePath returns the base configuration path for speaking to the server API.
+func basePath(configURL string) (charmhubpath.Path, error) {
+	baseURL := strings.TrimRight(configURL, "/")
+	rawURL := fmt.Sprintf("%s/%s", baseURL, path.Join("v2", "charms"))
```

Comment:
Maybe make "v2" and "charms" consts if we're removing them from config. Since this is the only place they're currently used, it seems reasonable not to require them in config.

Reply:
Yep, fair call -- I re-added them as constants (but non-exported this time).